automerge-wasm/src/observer.rs

```diff
@@ -0,0 +1,323 @@
+#![allow(dead_code)]
+
+use crate::interop::{export_value, js_set};
+use automerge::{ObjId, OpObserver, Parents, Prop, Value};
+use js_sys::{Array, Object};
+use wasm_bindgen::prelude::*;
+
+#[derive(Debug, Clone, Default)]
+pub(crate) struct Observer {
+    enabled: bool,
+    patches: Vec<Patch>,
+}
+
+impl Observer {
+    pub(crate) fn take_patches(&mut self) -> Vec<Patch> {
+        std::mem::take(&mut self.patches)
+    }
+    pub(crate) fn enable(&mut self, enable: bool) {
+        if self.enabled && !enable {
+            self.patches.truncate(0)
+        }
+        self.enabled = enable;
+    }
+
+    fn push(&mut self, patch: Patch) {
+        if let Some(tail) = self.patches.last_mut() {
+            if let Some(p) = tail.merge(patch) {
+                self.patches.push(p)
+            }
```

Comment:
This was quite confusing to read - I thought for a while that there was a missing `else`. It might read better if `Patch::merge` was called `Patch::attempt_merge`?